src/compiler/checker.ts

```diff
@@ -23563,12 +23565,40 @@ namespace ts {
             // and the type of the node includes type variables with constraints that are nullable, we fetch the
             // apparent type of the node *before* performing control flow analysis such that narrowings apply to
             // the constraint type.
-            if (type && isConstraintPosition(node) && forEachType(type, typeHasNullableConstraint)) {
+            if (type && isConstraintPosition(node) && someType(type, typeHasNullableConstraint)) {
                 return mapType(getWidenedType(type), getBaseConstraintOrType);
             }
             return type;
         }
 
+        function isTypeVariableWithUnionConstraint(type: Type) {
+            return !!(type.flags & TypeFlags.Instantiable && getBaseConstraintOrType(type).flags & TypeFlags.Union);
+        }
+
+        function containsTypeVariable(type: Type): boolean {
+            return !!(type.flags & TypeFlags.Instantiable || type.flags & TypeFlags.UnionOrIntersection && some((<UnionOrIntersectionType>type).types, containsTypeVariable));
+        }
+
+        function hasContextualTypeWithNoTypeVariables(node: Expression) {
+            const contextualType = !((isJsxOpeningElement(node.parent) || isJsxSelfClosingElement(node.parent)) && node.parent.tagName === node) && getContextualType(node);
+            return contextualType && !someType(contextualType, containsTypeVariable);
+        }
+
+        function getConstraintForReference(type: Type, reference: Identifier | ElementAccessExpression | PropertyAccessExpression | QualifiedName, checkMode: CheckMode | undefined) {

```

Comment:
what's the difference between a location and a reference? If it's really getConstraintOld and getConstraintNew, maybe this should be combined into one function.

Reply:
Yeah, not much of a difference really. I've merged it all into one function and renamed it to `getNarrowableTypeForReference`. Also renamed some of the helpers to better reflect what they do.